### DISCUSSION TITLE
gRPC Rust client

### DISCUSSION THREAD

**Discussion by gar1t** [GreptimeTeam/greptimedb#4644] [Q&A]
I'm building a Rust based data interface that uses a single instance of Greptime standalone. Standalone is running in its own process. I'd like to use a gRPC client from Rust. I'm looking at [the Rust ingester example](https://github.com/GreptimeTeam/greptimedb-ingester-rust/blob/master/examples/ingest.rs) but it looks like that client is configured for write-only (makes sense for ingest only). There's the "client" crate from the main greptimedb project, which appears to be a full blown client. I'd be incline to use that but the crate doesn't seem readily usable as an external library. This may be wrong though - I'm new to Rust!

Given my situation - writing a Rust app that wants to both write to and query a standalone Greptime instance over a socket as efficiently as possible - what would you recommend? I understand there are myriad interfaces available to me (MySQL, Postgres, etc.) but as I'm in Rust already, I'd prefer to use whatever Greptime itself uses when efficiency is of concern.

Any advice is greatly appreciated!

**Comment by v0y4g3r** [comment]
> I'm looking at [the Rust ingester example](https://github.com/GreptimeTeam/greptimedb-ingester-rust/blob/master/examples/ingest.rs) but it looks like that client is configured for write-only (makes sense for ingest only)

Yes, greptime-ingester-rust is meant to be write-only, as its name suggests

> There's the "client" crate from the main greptimedb project

This crate provides rpc interface and stubs for internal communication, eg., from frontend to datanode, and is not exposed to users.

> writing a Rust app that wants to both write to and query a standalone Greptime instance over a socket as efficiently as possible

- For efficient writing, I recommend using greptime-ingester-rust, a highly optimized data ingestion tool based on gRPC. We also offer an enterprise version of the data ingestion kit that uses shared memory IPC, which saves more CPU resources if the SDK is on the same host as the GreptimeDB server.
- For queries, there's no solution that allows writing and querying over the same socket. However, you can try [sqlx](https://docs.rs/sqlx/latest/sqlx/), or consider using [the MySQL line protocol implementation in native Rust](https://docs.rs/mysql/latest/mysql/) if an ORM seems overkill.

**Reply by gar1t** [reply]
This is surprising - that the gRPC server can't be used for queries. Looking over the proto defs, the server supports query defs. There's not a lot of support docs or examples (that I've seen anyway) to show a Rust client interacting with the gRPC server. I'll dig into your test code to look for Rust based interactions with a standalone server. It'd be nice to have a single interface in Rust, rather than go through a MySQL interface. I'll spend some time with this and report back here with any findings. Thanks for your help!

**Reply by sunng87** [reply]
@gar1t the reason we don't have query on our custom grpc protocol is we don't see much performance or features added when comparing with those database protocols. If your ingestion rate is not that high, it's possible to just use the database client and run insert statement for data ingestion. But if you have a fair amount of data, the grpc based ingestion sdk is preferred. 

Also I'm curious about your use-case. Could you provide more details about the application? You can share with me via email (sunning@greptime.com) if it's not for public.
